### ISSUE TITLE
Error when moving to downloaded directory

### ISSUE THREAD

**Issue by xbtowbw** [rakshasa/rtorrent#362]
We move completed torrentfolders from incoming directory to some downloaded-directory using this in .rtorrent.rc:

directory = /path/to/incoming

system.method.set_key = event.download.finished,move_complete,"d.set_directory=/path/to/downloaded/directory/;execute=mv,-n,$d.get_base_path=,/path/to/downloaded/directory/"

In most cases this works fine. But in some cases rtorrent gets stuck in some eternal loop. In the log (inside application when pressing l) we see about every 3 seconds:

( 8:47:14) F17F5821CF563AE10CD23BB2A4D08D66716F6F57->file_list: Failed to prepare file list: Could not create directory '/path/to/downloaded/directory/torrentfolder': Resource temporarily unavailablecreate directory '/path/to/downloaded/directory/torrentfolder': Resource temporarily unavailable

( 8:47:17) F17F5821CF563AE10CD23BB2A4D08D66716F6F57->file_list: Failed to prepare file list: Could not create directory '/path/to/downloaded/directory/torrentfolder': Resource temporarily unavailablecreate directory '/path/to/downloaded/directory/torrentfolder': Resource temporarily unavailable

( 8:47:20) F17F5821CF563AE10CD23BB2A4D08D66716F6F57->file_list: Failed to prepare file list: Could not create directory '/path/to/downloaded/directory/torrentfolder': Resource temporarily unavailablecreate directory '/path/to/downloaded/directory/torrentfolder': Resource temporarily unavailable

( 8:47:23) F17F5821CF563AE10CD23BB2A4D08D66716F6F57->file_list: Failed to prepare file list: Could not create directory '/path/to/downloaded/directory/torrentfolder': Resource temporarily unavailablecreate directory '/path/to/downloaded/directory/torrentfolder': Resource temporarily unavailable

This loops forever and rtorrent is blocked. Torrentfolder does not get moved from /path/to/incoming/torrentfolder to /path/to/downloaded/directory/torrentfolder. System just creates an empty /path/to/downloaded/directory/torrentfolder and prints the error message above forever.

This mostly happens with big torrents (200GB+)

Seems to be a bug. What can be done?

**Comment by chros73** [comment]
What OS do you use? (I'm on Ubuntu 14.04 .)
I'm using the `-f` switch (not `-n`) for `mv` but I haven't had any problem with moving directories, not even with large ones (100-600GB).

**Comment by xbtowbw** [comment]
Hi chros73
First of all: Your commands seem to correspond to some quite old rtorrent. What version are you using? We use the newest 0.9.6/0.13.6.
We now use the -f switch and stop and start like this:

system.method.set_key = event.download.finished,move_complete,"d.set_directory=/path/to/downloaded/directory/;execute=mv,-f,$d.get_base_path=,/path/to/downloaded/directory/;d.stop=;d.start="

But still we get the error:

( 8:47:14) F17F5821CF563AE10CD23BB2A4D08D66716F6F57->file_list: Failed to prepare file list: Could not create directory '/path/to/downloaded/directory/torrentfolder': Resource temporarily unavailablecreate directory '/path/to/downloaded/directory/torrentfolder': Resource temporarily unavailable

What we noticed is that this happens only on clients where rtorrent has used more than 50% of physical RAM. What could be done here? Could we somehow limit the ram rtorrent is using?

We tried max_memory_usage = 419430400 and also ulimit but both approaches do not limit the ram used...

What else could be done? Why does rtorrent needs again as much RAM as its already needed just to do a move in the filesystem?

**Comment by chros73** [comment]
I'm also using 0.9.6 (not from the master but the tag), and those commands aren't old at all, instead this one shouldn't work for you `system.method.set_key` unless you use some command line switch for rtorrent. The correct one is this: `method.set_key` .
Apart from that your command looks good.
What OS do you use?

For me RAM usage never goes above 800MB (there's an issue about this), try this (it sets to 2GB):
`max_memory_usage = 2000M`

**Comment by xbtowbw** [comment]
So how exactly and completely that line about method.set_key should look?

**Comment by pyroscope** [comment]
The question is: is that a cross-device move?

**Comment by xbtowbw** [comment]
Its not a cross-device move. Its all on the same partition. And we use Ubuntu 14.04.3 LTS.

**Comment by xbtowbw** [comment]
chros73 what exactly you mean with "d.directory.set=$d.custom=data_dir;execute=mv,-f,$d.base_path=,$d.custom=data_dir;d.stop=;d.start=" ? To add this additionally? To add this instead? But then were to specify the directory?

chros73 can you publish your full .rtorrent.rc ? Don't get the whole context here...

**Comment by chros73** [comment]
Yes, sorry, you're right, that line came from my full config: https://github.com/chros73/rtorrent-ps_setup/blob/master/ubuntu-14.04/home/chros73/.rtorrent.rc
Your line should look like this:
`method.set_key = event.download.finished,move_complete,"d.set_directory=/path/to/downloaded/directory;execute=mv,-f,$d.base_path=,/path/to/downloaded/directory;d.stop=;d.start="`

One more thing popped in my mind: isn't the base_path empty for those torrent? You can verify that with this:
`print=$d.base_path=`

If that's no that case, you can quickly try out rtorrent-ps build by PyroSope (if you're using x64 system) whether it result the same error or not: https://bintray.com/pyroscope/rtorrent-ps/rtorrent-ps#files (https://bintray.com/artifact/download/pyroscope/rtorrent-ps/rtorrent-ps_0.9.6-20150912-27aa4b7%7Etrusty_amd64.deb)
If not, then probably you have problem with your build.

**Comment by pyroscope** [comment]
the `d.stop=` belongs to the front btw, to avoid any race conditions.

**Comment by chros73** [comment]
@pyroscope yes, you're right, but strangely I haven't had any problem with this.

**Comment by chros73** [comment]
Actually, moving `d.stop=` to the beginning triggered a tracker controller bug in my setup (using that method with `download.finished` and `download.hash_done` events), so I have to use it the posted way.
More info: https://github.com/chros73/rtorrent-ps_setup/issues/9

**Comment by peondusud** [comment]
I tried chros73's code but when a download is complete I saw: 

_Below torrent line_
> Inactive: Download registered as completed, but hash check returned unfinished chunks.
and when I try to force start:

> 21A20DAC6A..0A4425B->file_list: Failed to prepare file list: Could not create directory 'cfg.download': Permission denied Could not create directory 'cfg.download': Permission denied

_and at rtorrent bottom:_
` Event 'event.download.finished' failed: Bad return code.`

I put a part of my config

```
method.insert = cfg.download,   private|const|string, (cat,(cfg.basedir),"download/")
method.insert = cfg.complete,   private|const|string, (cat,(cfg.basedir),"complete/")
directory.default.set = (cat,(cfg.basedir),"download")

schedule = watch_start,10,10,((load.start,(cat,(cfg.watch),"start/*.torrent")))
schedule = watch_load,15,10,((load.normal,(cat,(cfg.watch),"load/*.torrent")))
method.set_key = event.download.finished,move_complete,"d.directory.set=cfg.download;execute=mv,-f,$d.base_path=,cfg.complete;d.stop=;d.start="
```

**Comment by chros73** [comment]
Where's the `directory.default` comes from? It should be `$cfg.complete=`.

**Comment by peondusud** [comment]
I misunderstood I assumed it was  : 
`d.directory=old_folder ; mv old_folder  complete_folder`
Thanks now it works with :
`method.set_key = event.download.finished,move_complete,"d.directory.set=$cfg.complete=;execute=mv,-u,$d.base_path=,$cfg.complete=;d.stop=;d.start="`

But for already downloaded torrent in the old folder to force a move I must use ` event.download.hash_done` ?

**Comment by chros73** [comment]
That's right, but you'll need 1 more check in this case:
- whether the torrent is complete and only then apply what you want.

Reason behind this: if rtorrent starts (e.g. after a reboot) this download.hash_done event is triggered for every torrent.

If you like the idea of moving data take a look at this config, it's full with crazy but useful ideas: https://github.com/chros73/rtorrent-ps_setup/blob/master/ubuntu-14.04/home/chros73/.rtorrent.rc

**Comment by peondusud** [comment]
Thanks a lot. I saw some posts related to pyroscope-rtorrentPS-pryocore
What's the pros & cons?
I try to create a pseudo docker torrent image with all goods. Pyroscope wiil be usefull if I already use rutorrent(web)

**Comment by peondusud** [comment]
Now I have an issue when I try to remove the torrent.
`Event 'event.download.erased' failed: Bad return code.`

I try to fix
`schedule = untied_torrents, 8, 5, "remove_untied="`

or

```
schedule = untied_directory, 10, 10, stop_untied=
schedule = untied_directory, 10, 10, close_untied=
schedule = untied_directory, 10, 10, remove_untied=
```
Always have  the same error.

**Comment by chros73** [comment]
How do you delete them? What's the related config? (use pastebin if it's long)
This line `schedule = untied_torrents, 8, 5, "remove_untied="` does this only: Removes torrents from client when its metafile (torrent file) has been deleted manually or by a script. Run it every 5 seconds.
The error caused by the actual `event.download.erased` event.

"I saw some posts related to pyroscope-rtorrentPS-pryocore."
It's simply amazing. After you got used to it you don't need any additional UI (GUI, webUI, etc) for rtorrent. I even patched it to be able to show all the info that I use: https://github.com/chros73/rtorrent-ps#fork-notes

**Comment by peondusud** [comment]
The desired file
https://github.com/peondusud/Dockers-/blob/master/debian/rtorrent%2Brutorrent/rtorrent.rc

> you don't need any additional UI (GUI, webUI, etc) for rtorrent. 
rtorrent is good for desktop. pyro is supposed to be an rtorrent enhancer. How it deals with remote 
 control like rutorrent do.

**Comment by chros73** [comment]
There's no `event.download.erased` event defined in that config. Never was?

"pyro is supposed to be an rtorrent enhancer. How it deals with remote control like rutorrent do."
That's right, that's pyrocore. But I have been talking about rtorrent-ps! See above.

**Comment by peondusud** [comment]
> There's no event.download.erased event defined in that config. Never was?
Yes I never used this event.

**Comment by chros73** [comment]
In this case I don't know, I don't see anything unusual in your config. Try to raise the log level maybe you can see something in it.

Btw, I think you only need just 1 of these (if you want to use it at all):

```
schedule = untied_torrents, 8, 5, "remove_untied="
schedule = untied_directory, 10, 10, stop_untied=
schedule = untied_directory, 10, 10, close_untied=
schedule = untied_directory, 10, 10, remove_untied=
```

**Comment by peondusud** [comment]
With 
```
schedule = untied_torrents, 8, 5, "remove_untied="
schedule = untied_directory, 10, 10, stop_untied=
schedule = untied_directory, 10, 10, close_untied=
schedule = untied_directory, 10, 10, remove_untied=
```

In log when I try form rutorrent to force delete torrent and linked data
```
1464815058 I 0000000000000000000000000000000000000000->chunk_list: Clearing.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->download_list: Erasing download.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->download_list: Closing download with throw.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->download_list: Pausing download: flags:0.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->download: Stopping torrent: flags:0.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->tracker_controller: Sending stopped event.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->tracker_list: Sending 'stopped' to group:0 url:'http://open.nyaatorrents.info:6544/announce'.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->tracker_controller: Called disable with 1 trackers.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->download: Closing torrent: flags:0.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->hash_torrent: Clear.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->file_list: Closing.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->chunk_list: Clearing.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->download_list: Opening download.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->download: Opening torrent: flags:ffffffff.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->file_list: Opening.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->chunk_list: Resizing: from:0 to:496.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->download_list: Opening download.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->tracker_list: Received 0 peers from tracker url:'http://open.nyaatorrents.info:6544/announce'.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->download_list: Pausing download: flags:0.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->hash_torrent: Clear.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->file_list: Closing.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->chunk_list: Clearing.
1464815700 I 8BB35A6F7614095578F63F9D5A0769174AC14DB4->hash_torrent: Clear.
```

**Comment by pyroscope** [comment]
Decide who's the master, and who deletes data. If you jump into the middle of ruT actions with some event with side effects, chaos is the obvious consequence.
